src/appmixer/utils/csv/CSVProcessor.js

```diff
@@ -367,34 +367,78 @@ module.exports = class CSVProcessor {
      * @return {Promise<*>}
      * @public
      */
-    async addRow(newRow, closure) {
+    async addRow({ row, rowWithColumns }, closure) {
 
-        const stream = await this.loadFile();
+        const config = this.context.config;
+        const lockOptions = {
+            ttl: parseInt(config.lockTTL, 10) || 60000, // Default 1 minute TTL
+            retryDelay: 500
+        };
+        const lock = await this.context.lock(this.fileId, lockOptions);
 
-        let idx = 0;
-        const writeStream = new PassThrough();
-        stream.on('data', (row) => {
-            try {
-                writeStream.write(row.join(this.delimiter) + '\n');
-                if (closure(idx, row, false)) {
-                    writeStream.write(newRow.join(this.delimiter) + '\n');
-                }
-                idx += 1;
-            } catch (err) {
-                stream.destroy(err);
-            }
-        }).on('error', (err) => {
-            stream.destroy(err);
-        }).on('end', async () => {
-            if (closure(idx, null, true)) {
-                writeStream.write(newRow.join(this.delimiter) + '\n');
+        try {
+            const stream = await this.loadFile();
+            await this.loadHeaders();
+
+            let rowAsArray;
+
+            // Process row data based on whether headers are included
+            if (this.withHeaders) {
+                const headers = this.getHeaders();
+                const parsed = expressionTransformer(rowWithColumns);
+                rowAsArray = headers.map(() => ''); // Initialize rowAsArray with empty strings
+                parsed.forEach(({ column, value }) => {
+                    const idx = this.getHeaderIndex(column);
+                    if (idx !== -1) { // Ensure the column exists in headers
+                        rowAsArray[idx] = value;
+                    }
+                });
+            } else {
+                rowAsArray = row.split(this.delimiter);
             }
-            writeStream.end();
-        });
 
-        return await this.context.replaceFileStream(this.fileId, writeStream);
+            // Ensure each item in rowAsArray is not undefined or null
+            rowAsArray = rowAsArray.map(item => item ?? '');
+
+            let idx = 0;
+            const writeStream = new PassThrough();
+
+            stream.on('data', (rowData) => {
+                writeStream.write(rowData.join(this.delimiter) + '\n');
+                if (closure(idx, rowData, false)) {
+                    writeStream.write(rowAsArray.join(this.delimiter) + '\n');
+                }
+                idx++;
+
+                // Extend lock periodically
+                if (idx % 1000 === 0) {
```

Comment:
<img width="1919" alt="Discover_-_Elastic" src="https://github.com/clientIO/appmixer-connectors/assets/928772/a40f172c-02ff-4a37-bd06-acf787416808">
this code generated over 56K of logs.

Reply:
Should extend based on a `time` rather the row count, that is too often.